### PR TITLE
Gr100 uses items not in marketplaces

### PR DESCRIPTION
**Pull request by barryyosi-panw**
## Related Issues
fixes: [CIAC-10572 ](https://jira-dc.paloaltonetworks.com/browse/CIAC-10572)

## Description
Refactoring 'marketplaces' field validation (Content item uses items not in marketplace - GR100)

### CODE REVIEW

File: demisto_sdk/commands/validate/validators/GR_validators/GR100_uses_items_not_in_market_place.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+from demisto_sdk.commands.content_graph.objects.parsing_rule import ParsingRule
+from demisto_sdk.commands.content_graph.objects.playbook import Playbook
+from demisto_sdk.commands.content_graph.objects.report import Report
+from demisto_sdk.commands.content_graph.objects.script import Script
+from demisto_sdk.commands.content_graph.objects.test_playbook import TestPlaybook
+from demisto_sdk.commands.content_graph.objects.trigger import Trigger
+from demisto_sdk.commands.content_graph.objects.widget import Widget
+from demisto_sdk.commands.content_graph.objects.wizard import Wizard
+from demisto_sdk.commands.content_graph.objects.xsiam_dashboard import XSIAMDashboard
+from demisto_sdk.commands.content_graph.objects.xsiam_report import XSIAMReport
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/validators/GR_validators/GR100_uses_items_not_in_market_place.py:37]: I think you can switch it with basecontent (see BA114 as a reference).

File: demisto_sdk/commands/validate/validators/GR_validators/GR100_uses_items_not_in_market_place.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    XSIAMDashboard,
+    XSIAMReport,
+]
+
+
+class MarketplacesFieldValidator(BaseValidator[ContentTypes]):
+    error_code = "GR100"
+    description = (
+        "Content item attempts to use other content items which are not supported in all of the "
+        "marketplaces of the content item."
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/validators/GR_validators/GR100_uses_items_not_in_market_place.py:80]:
```suggestion
        "Detect content items that attempt to use other content items which are not supported in all of the "
        "marketplaces of the content item."
```

File: demisto_sdk/commands/validate/validators/GR_validators/GR100_uses_items_not_in_market_place.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    )
+    rationale = "Content graph proper construction."
+    error_message = ""
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/validators/GR_validators/GR100_uses_items_not_in_market_place.py:83]: ?

File: demisto_sdk/commands/validate/validators/GR_validators/GR100_uses_items_not_in_market_place.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    related_field = "marketplaces"
+    is_auto_fixable = False
+    expected_git_statuses = [
+        GitStatuses.RENAMED,
+        GitStatuses.ADDED,
+        GitStatuses.MODIFIED,
+    ]
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/validators/GR_validators/GR100_uses_items_not_in_market_place.py:90]: Let's remove this, putting it under use_git is sufficient.

File: demisto_sdk/commands/validate/validators/GR_validators/GR100_uses_items_not_in_market_place.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        ):
+
+            used_content_items = [
+                item.content_item_to.object_id
+                for item in content_item.relationships_data.get(RelationshipType.USES)
+            ]
+
+            error_message, _ = Errors.uses_items_not_in_marketplaces(
+                content_item.name, content_item.marketplaces, used_content_items
+            )
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/validators/GR_validators/GR100_uses_items_not_in_market_place.py:106]: We don't want to use the old errors.
you should copy/alter the message and put it under error_message.
If you wish to send any variables use format (see other validations as a reference).

File: demisto_sdk/commands/validate/tests/GR_validators_test.py [hunk 6]
```diff
@@ -0,0 +1,508 @@
...
+    - Validate the existence of invalid marketplaces usages.
+    """
+
+    graph_interface = ContentGraphInterface()
+    create_content_graph(graph_interface)
+    MarketplacesFieldValidator.graph_interface = graph_interface
+    validation_results = MarketplacesFieldValidator().is_valid(repository.packs)
+    validation_results_messages = {
+        validation_result.message for validation_result in validation_results
+    }
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/tests/GR_validators_test.py:502]: Let's validate using 'in'.
Also, let's validate the expected length of the result.

File: demisto_sdk/commands/validate/tests/GR_validators_test.py [hunk 7]
```diff
@@ -0,0 +1,508 @@
...
+
+def test_MarketplacesFieldValidator_is_valid(repository: ContentDTO):
+    """
+    Given
+    - A content repo
+    When
+    - running MarketplacesFieldValidator is_valid function.
+    Then
+    - Validate the existence of invalid marketplaces usages.
+    """
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/tests/GR_validators_test.py:494]: Try to explain the case more thoroughly - what exactly points to what, what should fail and what shouldn't

**Review by YuvHayun — CHANGES_REQUESTED** [review]
Great work!
see my comments.

File: demisto_sdk/commands/validate/validators/GR_validators/GR100_uses_items_not_in_market_place.py (hunk 4 above)
Comment (barryyosi-panw) [reply demisto_sdk/commands/validate/validators/GR_validators/GR100_uses_items_not_in_market_place.py]: done

File: demisto_sdk/commands/validate/validators/GR_validators/GR100_uses_items_not_in_market_place.py (hunk 5 above)
Comment (barryyosi-panw) [reply demisto_sdk/commands/validate/validators/GR_validators/GR100_uses_items_not_in_market_place.py]: adapted the error message construction.

File: demisto_sdk/commands/validate/tests/GR_validators_test.py (hunk 6 above)
Comment (barryyosi-panw) [reply demisto_sdk/commands/validate/tests/GR_validators_test.py]: done

File: demisto_sdk/commands/validate/tests/GR_validators_test.py (hunk 7 above)
Comment (barryyosi-panw) [reply demisto_sdk/commands/validate/tests/GR_validators_test.py]: Done

File: demisto_sdk/commands/validate/validators/GR_validators/GR100_uses_items_not_in_market_place.py (hunk 3 above)
Comment (barryyosi-panw) [reply demisto_sdk/commands/validate/validators/GR_validators/GR100_uses_items_not_in_market_place.py]: adapted it

File: demisto_sdk/commands/validate/validators/GR_validators/GR100_uses_items_not_in_market_place.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    )
+
+    related_field = "marketplaces"
+    is_auto_fixable = False
+
+    def is_valid(self, content_items: Iterable[ContentTypes]) -> List[ValidationResult]:
+        validation_results = []
+
+        for content_item in self.graph.find_uses_paths_with_invalid_marketplaces(
+            [item.pack_id for item in content_items]
```
Comment (dantavori) [inline demisto_sdk/commands/validate/validators/GR_validators/GR100_uses_items_not_in_market_place.py:94]: need to discuss this
@moishce @YuvHayun @ShahafBenYakir FYI

File: demisto_sdk/commands/validate/tests/test_graph_tools.py
Comment (barryyosi-panw) [inline demisto_sdk/commands/validate/tests/test_graph_tools.py:1]: [reference](https://github.com/demisto/demisto-sdk/blob/master/demisto_sdk/commands/content_graph/tests/graph_validator_test.py)

**Comment by barryyosi-panw** [comment]
Adapted the validation to support the new execution modes

File: demisto_sdk/commands/validate/validators/GR_validators/GR100_uses_items_not_in_market_place_all_files.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    TestPlaybook,
+    Trigger,
+    Widget,
+    GenericDefinition,
+    GenericField,
+    GenericModule,
+    GenericType,
+    XSIAMDashboard,
+    XSIAMReport,
+]
```
[the remark this answers is not in the document]
Comment (barryyosi-panw) [reply demisto_sdk/commands/validate/validators/GR_validators/GR100_uses_items_not_in_market_place_all_files.py]: done

**Review by YuvHayun — APPROVED** [review]
Great job!
Approved from my side.

**Review by thefrieddan1 — APPROVED** [review]

**Pull request merged**